File: InfoQ/14.py
<html><head><meta http-equiv="content-type" content="text/html; charset=utf-8" /></head><body><h3>Angular.JS と ASP.NET を利用した Single Page アプリケーションの作成</h3><p><a target="_blank" href="http://www.infoq.com/news/2014/04/asp_net-angular"><em>原文(投稿日：2014/04/04)へのリンク</em></a></p>
<div class="article_page_left news_container text_content_container"> 
 <div class="text_info"> 
  <p class="MsoNormal">Build の水曜日セッションにて、David Catuhe 氏と Jon Galloway 氏が <a href="http://angularjs.org/">AngularJS</a> を ASP.NET アプリケーションで利用する方法をプレゼンテーションした。これにより、開発者はモダンな Single Page アプリケーションを高速に作成する方法を提供された。</p> 
  <p class="MsoNormal">Angular は Google によって作成され、現在はオープンソースプロジェクトとして扱われてサポートされている。名前の示す通り JavaScript ベースのライブラリであり、Model View Controller (MVC) 設計パターンにしたがっている。Catuhe 氏と Galloway 氏の説明する通り、Angular は DI（依存注入）を利用して ASP.NET アプリケーションを機能強化する。単一ファイルである <code>angular.min.js</code> は、機能を有効化するために一度だけ読み込む必要がある。(NuGet ユーザは最新版（β）か安定板を <a href="http://www.nuget.org/packages/angularjs/">取得</a> することができる）。</p> 
  <p class="MsoNormal">Angular の利用方法としては、全てに適用／全てに適用しないという二者択一のアプローチではないことを指摘しており、作成しているサイトに対し選択要素の機能のみを利用するといった方法も考えられる。 どのイベントでも Catuhe 氏と Galloway 氏は Scripts フォルダの下に “apps” を作成して体系化することを推奨している。</p> 
  <p class="MsoNormal">JavaScript の圧縮についての重要な警告：標準で利用される圧縮の場合、コード内の Angular における依存注入を破壊する恐れがある（Angular のチュートリアルドキュメントである <a href="http://docs.angularjs.org/tutorial/step_05">XHRs &amp; Dependency Injection</a> の“A Note on Minification”に記載されている）。</p> 
  <p class="MsoNormal">Angular を自身のサイト内で利用する場合、html タグ内に “ng-app” を追加する必要がある。</p> 
  <p><code> </code></p> 
  <p class="MsoNormal"><code>&lt;html ng-app … &gt;
    <o:p></o:p></code></p> 
  <p><code> </code></p> 
  <p>上記を記載することが Angular の実行準備となる。Angular は <code>$http</code> を利用し、jQuery の軽量バージョンのファイルをロードする。すでに jQuery を配置して利用している場合、Angular は一貫性を維持するために配置済みの jQuery を利用する。
   <o:p></o:p>
   <o:p>
    &nbsp;
   </o:p></p> 
  <p class="MsoNormal">Catuhe 氏と Galloway 氏はマジック：ザ・ギャザリング のカードのサンプルアプリケーションを利用し、Single Page アプリケーション（SPA）ベースでの表示／情報格納についてデモを実施した。UI 構築時に SPA はビューを利用するが、Angular 自身はそれらのビューで定義されたルーティングを利用する。</p> 
  <p class="MsoNormal">Angular で MVC Route を用いる際にディープリンクが衝突する場合があるので、彼らは Catch All Route の利用を提案している。
   <o:p></o:p></p> 
  <p>&nbsp;</p> 
  <blockquote>
   <code> <p class="MsoNormal">routes.MapRoute(<br /> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; name: &quot;Catch all route for SPA&quot;,<br /> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; url: &quot;App/{*catchall}&quot;,<br /> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; defaults: new{<br /> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; controller = &quot;Home&quot;, <br /> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; action = &quot;Index&quot;});
     <o:p></o:p></p> </code>
  </blockquote> 
  <p>&nbsp;</p> 
  <p class="MsoNormal">HTML表示における他のTIPSが存在する。ビューから生成されたHTMLの場合は複雑な処理は必要な一方で、単一のファイル（MyHTML.html）からなる HTML の場合は IIS Rewrite ルールを利用して URL を変更することだ。
   <o:p></o:p></p> 
  <p><code> </code></p> 
  <p class="MsoNormal"><code>/myHTML.html -&gt; /myHTML
    <o:p></o:p></code></p> 
  <code> </code> 
  <p>&nbsp;</p> 
  <p class="MsoNormal">彼らのデモアプリケーションの動作を閲覧したい場合、<a href="http://channel9.msdn.com/Events/Build/2014/3-644">Channel9 ページ</a> に情報提供を相談して欲しい。
   <o:p></o:p></p> 
  <p>&nbsp;</p> 
 </div> 
</div><br><br><br><br><br><br></body></html>

File: InfoQ/14.java
<html><head><meta http-equiv="content-type" content="text/html; charset=utf-8" /></head><body><h3>Facebook Unveils Jest for JavaScript Unit Testing, Automatic Mocking</h3><p>Facebook has released <a href="http://facebook.github.io/jest/">Jest</a>, an open source&nbsp;Unit Testing tool for JavaScript, built on top of <a href="http://jasmine.github.io/">Jasmine</a>.</p>
<p>Jest was originally conceived by Facebook two years ago as a way to solve the need for writing fast, reliable tests for the web chat application. As it gained interest internally the project was picked up six months ago by Jeff Morrison, a software engineer for Facebook, who set out to make performance improvements and take Jest open source.</p>
<p>At its most basic level, Jest is designed to make it faster and easier to write idiomatic JavaScript tests. Jest automatically mocks <a href="http://www.commonjs.org/">CommonJS</a> modules returned by require(), and other features include built-in support for DOM APIs in test environments, sensible defaults, support for pre-processing code, and parallel test execution by default.&nbsp;By running tests concurrently in parallel processes, Jest has test runs finish sooner.</p>
<p>Morrison said:</p>
<blockquote> 
 <p>The goal for Jest is to reduce the time spent and cognitive load necessary to get started with testing a project -- and so it provides most of the things you need out of the box: A fast CLI, a set of mocking utilities, and its automatic module mocking system.</p> 
 <p>Additionally, if you're looking for isolation tools like mocking libraries, most other tools will leave you writing quite an unsatisfactory bit of boilerplate in your tests (and often even in your main code) just to get going.</p> 
 <p>We've seen first-hand at Facebook how important it is to spend more time writing your app (vs time spent getting set up to write your app) -- and so this is the problem Jest is focused on solving.</p> 
</blockquote>
<p>Jest differs from the Jasmine framework by adding several layers on top. Most notably, Jest automatically mocks dependencies when running tests. Jest automatically generates mocks for each of a module's depenedencies, and provides those mock by default -- making it easier to&nbsp;isolate a module from its dependencies. Morrison said that isolation is the default for new tests, and developers would now have &quot;complete control&quot; over how much isolation is needed. Each test indicates which modules should or should not be mocked.</p>
<p>Facebook's documentation goes into <a href="http://facebook.github.io/jest/docs/automatic-mocking.html">further detail</a> on Automatic Mocking:</p>
<blockquote>
 Jest actually implements its own version of the require() function in the testing environment. Jest's custom require() function loads the real module, inspects what it looks like, and then makes a mocked version based on what it saw and returns that. 
 <p>This means Jest is going to give you an object with the same shape as the real module, but with mocks for each of the exported values instead of the real values.</p> 
</blockquote>
<p>Although Jest introduces automatic mocking, it is also noted that it is still possible for developers to control what is and is not mocked by providing jest.mock() and jest.dontMock() APIs for customization.</p>
<p>Reaction from the community was mostly positive. On Hacker News, user Cthulu <a href="https://news.ycombinator.com/item?id=7748886">said</a>:</p>
<blockquote>
 Looks interesting: the test suite for our current AngularJS project is slowly slowing down, in part because there's just more tests, but the major performance bottlenecks are: 
 <ul> 
  <li>No parallelisation, even if test suites are all independent</li> 
  <li>DOM tests, which cause a lot of GC pauses</li> 
  <li>(Probably) PhantomJS startup and initialisation (not measured)</li> 
 </ul> I've done a simple optimization where my tests get split in the middle and run in two separate terminals (during development, continuous testing), but it's kinda iffy.
 <br /> 
</blockquote>
<p>Directly addressing dependency injection and AngularJS, Facebook <a href="http://facebook.github.io/jest/docs/common-js-testing.html#content">says</a>&nbsp;&quot;Jest&nbsp;achieves the same result using a different approach.&quot; With Angular, code is written by passing dependencies as arguments, making it very easy to write a test. However, Facebook notes that in order to have a testable function in Angular, developers conform to its specific pattern and pass it into Angular's dependency injection framework. The solution in Jest is slightly different:</p>
<blockquote>
 Jest allows for mocking dependencies in the same way that Angular does, but instead of building a proprietary module loader, it uses CommonJS. This enables you to test any existing code that already uses CommonJS without having to heavily refactor it to make it compatible with another module system.
 <br /> 
</blockquote>
<p>User Caiob agreed with the positive sentiment about Jest and was a fan of the approach to dependency injection,&nbsp;<a href="https://news.ycombinator.com/item?id=7745011">saying</a>: &quot;It's great that Facebook are improving an existing/familiar tool like Jasmine. Also, I love the way they treat CommonJS modules.&quot;</p>
<p>With Jest, Morrison says Facebook want to start of a trend toward making testing easy, leaving developers more time to writing their app. InfoQ readers who would like to contribute to the project can check out the<a href="http://github.com/facebook/jest"> github repo </a>and send pull requests, or join in #jestjs on freenode.</p><br><br><br><br><br><br></body></html>